Invalid memory operations on calling initialize/destroy functions repetitively

There's a bug when utilizing dcurl after first time. Calling `dcurl_destroy` in second time would yield double free error:

```
[dcurl] Implementation CPU (Intel SSE) is initialized successfully
test-dcurl(34354,0x1115395c0) malloc: *** error for object 0x7f9ed4d00000: pointer being freed was not allocated
test-dcurl(34354,0x1115395c0) malloc: *** set a breakpoint in malloc_error_break to debug
/bin/sh: line 1: 34354 Abort trap: 6           ./build/test-dcurl
make: *** [build/test-dcurl.done] Error 134
```

Step to reproduce:
Put another `dcurl_init` and `dcurl_destroy` in `test_dcurl` test, then simply `make check`.

--- a/src/dcurl.c
+++ b/src/dcurl.c
@@ -94,6 +94,7 @@ void dcurl_destroy()
     list_for_each(p, &IMPL_LIST) {
         impl = list_entry(p, ImplContext, list);
         destroyImplContext(impl);
+        list_del(p);
     }
 }
 

--- a/tests/test-dcurl.c
+++ b/tests/test-dcurl.c
@@ -2,6 +2,8 @@
 #include "common.h"
 #include "dcurl.h"
 
+#define LOOP_MAX 5
+
 int main()
 {
     char *trytes =
@@ -47,27 +49,29 @@ int main()
 
     int mwm = 9;
 
-    /* test dcurl Implementation with mwm = 9 */
-    dcurl_init();
-    int8_t *ret_trytes = dcurl_entry((int8_t *) trytes, mwm, 0);
-    assert(ret_trytes);
-    dcurl_destroy();
+    for (int loop_count = 0; loop_count < LOOP_MAX; loop_count++) {
+        /* test dcurl Implementation with mwm = 9 */
+        dcurl_init();
+        int8_t *ret_trytes = dcurl_entry((int8_t *) trytes, mwm, 0);
+        assert(ret_trytes);
+        dcurl_destroy();
 
-    Trytes_t *trytes_t = initTrytes(ret_trytes, 2673);
-    assert(trytes_t);
-    Trytes_t *hash_trytes = hashTrytes(trytes_t);
-    assert(hash_trytes);
+        Trytes_t *trytes_t = initTrytes(ret_trytes, 2673);
+        assert(trytes_t);
+        Trytes_t *hash_trytes = hashTrytes(trytes_t);
+        assert(hash_trytes);
 
-    /* Validation */
-    Trits_t *ret_trits = trits_from_trytes(hash_trytes);
-    for (int i = 243 - 1; i >= 243 - mwm; i--) {
-        assert(ret_trits->data[i] == 0);
+        /* Validation */
+        Trits_t *ret_trits = trits_from_trytes(hash_trytes);
+        for (int i = 243 - 1; i >= 243 - mwm; i--) {
+            assert(ret_trits->data[i] == 0);
+        }
+
+        free(ret_trytes);
+        freeTrobject(trytes_t);
+        freeTrobject(hash_trytes);
+        freeTrobject(ret_trits);
     }
-
-    free(ret_trytes);
-    freeTrobject(trytes_t);
-    freeTrobject(hash_trytes);
-    freeTrobject(ret_trits);
 
     return 0;
 }

